### PR TITLE
eclass/rpm-extended: add eclass

### PR DESCRIPTION
**Pull request by AndrewAmmerlaan**
Package-Manager: Portage-3.0.12, Repoman-3.0.2
Signed-off-by: Andrew Ammerlaan <andrewammerlaan@riseup.net>

### CODE REVIEW

**Comment by AndrewAmmerlaan** [comment]
I just now made one small change to the eclass, instead of doing `source ....sh` I now dow `bash ....sh`, this way emerge doesn't complain if the script calls exit, which means that I don't have to perform some sed magic to remove all exit commands.

**Comment by epsilon-0** [comment]
Without looking too much into the eclass yet, maybe rpm-extended isn't the best name?
Case in point, there is the intel-sdp-r1 eclass, which also does a bunch of rpm things, but the name is specific to make it clear it's doing intel specific rpm things. Maybe the name here should be reflective of the sci-ni things that are happening.
This is easily fixable so no need to change just yet, we can do this at the last part.

**Comment by AndrewAmmerlaan** [comment]
> Without looking too much into the eclass yet, maybe rpm-extended isn't the best name?
> Case in point, there is the intel-sdp-r1 eclass, which also does a bunch of rpm things, but the name is specific to make it clear it's doing intel specific rpm things. Maybe the name here should be reflective of the sci-ni things that are happening.
> This is easily fixable so no need to change just yet, we can do this at the last part.

In principle this eclass should work for *any* rpm file, all it does extra is extract the rpm's post/pre(un)install scripts and runs them, it is not really specific to rpm files from Nation Instruments, this is on purpose, I tried to write in such a way that it is universally applicable. I suppose I could still rename it to something like `auto-rpm.eclass`.

I suppose I could (re)write the eclass in such a way that _is_ specific to these NI ebuilds, in which case I could move some stuff that is now in the script that generates the ebuilds to the eclass, such as `LICENSE`, `KEYWORDS`, `pkg_nofetch`, but IMO it is better to keep eclases as universally as possible.

File: eclass/rpm-extended.eclass [hunk 1]
```diff
@@ -0,0 +1,76 @@
+# Copyright 1999-2020 Gentoo Foundation
+# Distributed under the terms of the GNU General Public License v2
+
```
Comment (epsilon-0) [inline eclass/rpm-extended.eclass:3]: For cross-reference, the python-single-r1.eclass is a good place to look at for my comments: https://gitweb.gentoo.org/repo/gentoo.git/tree/eclass/python-single-r1.eclass

File: eclass/rpm-extended.eclass [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+# Author : Andrew Ammerlaan <andrewammerlaan@riseup.net> (16-12-2020)
```
Comment (epsilon-0) [inline eclass/rpm-extended.eclass:4]: add `@AUTHOR`

File: eclass/rpm-extended.eclass (hunk 1 above)
Comment (epsilon-0) [inline eclass/rpm-extended.eclass:3]: add `@ECLASS`, `@MAINTAINER` and basically all the things from the header from the `python-single-r1` eclass.

File: eclass/rpm-extended.eclass [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+# post/pre(un)install scripts and runs them.
+#
+# This is mostly useful when creating ebuilds for rpm files in bulk.
+# In cases when it is not doable to write a Gentoo specific src_install
+# and pkg_post/pre/install/rm functions for each ebuild.
+
+inherit rpm
+
+# Need rpm to extact scripts
+BDEPEND="app-arch/rpm"
```
Comment (epsilon-0) [inline eclass/rpm-extended.eclass:18]: This should already be there from `inherit rpm`

File: eclass/rpm-extended.eclass [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+#
+# Extension of the rpm.eclass by Alastair Tse <liquidx@gentoo.org>
+#
+# This also installs *all* files in the rpm, and extracts
+# post/pre(un)install scripts and runs them.
+#
+# This is mostly useful when creating ebuilds for rpm files in bulk.
+# In cases when it is not doable to write a Gentoo specific src_install
+# and pkg_post/pre/install/rm functions for each ebuild.
+
```
Comment (epsilon-0) [inline eclass/rpm-extended.eclass:14]: add the supported `EAPI` checks.

File: eclass/rpm-extended.eclass [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+# Need rpm to extact scripts
+BDEPEND="app-arch/rpm"
+
+# Otherwise we get the S does not exist error
+S="${WORKDIR}"
+# Everything in the rpm is prebuilt
+QA_PREBUILT="*"
+
+rpm-extended_src_compile() {
```
Comment (epsilon-0) [inline eclass/rpm-extended.eclass:25]: maybe add a very small documentation header saying it does nothing like: https://gitweb.gentoo.org/repo/gentoo.git/tree/eclass/python-single-r1.eclass#n519

File: eclass/rpm-extended.eclass [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+	# Nothing to do here
+	return
+}
+
+rpm-extended_src_install() {
```
Comment (epsilon-0) [inline eclass/rpm-extended.eclass:30]: As I am unfamiliar with how RPMs work (and I am sure there might be others) can you add small documentations for what all the functions do?

**Review by epsilon-0 — CHANGES_REQUESTED** [review]
Mostly comments on adding documentation.

**Comment by AndrewAmmerlaan** [comment]
> Hmm, unfortunately, I am unfamiliar with how RPMs are packaged and what their structure is so I am unable to comment and check if this eclass works perfectly for all RPMs.
> I've asked @SoapGentoo and lets see if he has any comments.


In principle a RPM file is an archive with some metadata. The metadata contains thins like, name, version, revision, description, url and stuff, we can extract this information with `rpm --info`. This is basically what the script in the other PR is doing. Along with this metadata 5 scripts can be included, preinstall, postinstall, preuninstall, postuninstall and verify, we can view these scripts with `rpm --scripts`. Unfortunately there is no way (that I know) to have rpm output only the script for a specific stage, instead it outputs them all at once. Therefore we use some sed magic to separate each script from this output, write it to a file, and execute it. And this is basically the part that the eclass does. Together the ebuild-generation-script and the eclass, create something which is more or less like an "ebuild-wrapper" around the rpm file, which does the exact same thing that would happen if you would install the rpm file using `rpm --install` except now dependencies and files are tracked by portage, which is more user friendly than having to run two separate package managers (portage and rpm) on one system.

File: eclass/rpm-extended.eclass (hunk 6 above)
Comment (AndrewAmmerlaan) [reply eclass/rpm-extended.eclass:30]: On it

File: eclass/rpm-extended.eclass (hunk 3 above)
Comment (AndrewAmmerlaan) [reply eclass/rpm-extended.eclass:18]: The rpm eclass does not actually add a dependency on rpm, instead it adds a dependency on rpm2targz, see: https://gitweb.gentoo.org/repo/proj/prefix.git/tree/eclass/rpm.eclass?id=d7fc8cf65c536224bace1d22c0cd85a526490a1e

Comment (epsilon-0) [reply eclass/rpm-extended.eclass]: :open_mouth:

**Comment by AndrewAmmerlaan** [comment]
@epsilon-0 I added some documentation in the eclass, and improved the overall style of the thing to, let me know if that clarifies things. :)

File: eclass/rpm-extended.eclass [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+# and pkg_post/pre/install/rm functions for each ebuild.
+
+# The rpm eclass provides the pkg_unpack function we need
+inherit rpm
+
+case "${EAPI:-0}" in
+	0|1|2|3|4|5)
+		die "Unsupported EAPI=${EAPI:-0} (too old) for ${ECLASS}"
+		;;
+	6|7)
```
Comment (epsilon-0) [inline eclass/rpm-extended.eclass:27]: Do we need 6? We can simplify life by just keeping it 7 yea?

Comment (AndrewAmmerlaan) [reply eclass/rpm-extended.eclass:27]: Well no we don't really need 6, but from what I understand there are some eclasses that are not compatible with 7 yet (gnome-something if I remember correctly), by keeping 6 in we make sure this eclass is compatible with those eclasses.

Comment (epsilon-0) [reply eclass/rpm-extended.eclass:27]: I'd say, let's remove 6. Saves future work. 
Highly unlikely we are going to be crossing paths with gnome.

Comment (AndrewAmmerlaan) [reply eclass/rpm-extended.eclass]: Alright, 6 is gone :)

**Comment by epsilon-0** [comment]
OK, documentation wise lgtm.
Still going to wait for other comments.

**Comment by epsilon-0** [comment]
OK, this is not logically correct. pkg_* stages don't have access to the distfiles. What if package is being installed as a binpkg?

**Comment by AndrewAmmerlaan** [comment]
> OK, this is not logically correct. pkg_* stages don't have access to the distfiles. 

What do you mean? The variable DISTDIR exists in these phases and the rpm command can access and read the files in it successfully.

> What if package is being installed as a binpkg?

All these ebuilds are RESTRICT="mirror bindist" so there shouldn't be any binpkg's of these package's. That being said, perhaps RESTRICT="bindist" should be moved into the eclass instead of leaving it up to the ebuild to set it? Since distributing binaries of something that already is a binary does not make any sense anyway?

**Comment by epsilon-0** [comment]
Not sure if mirror restriction is needed but bindist is for sure. The pkg_* functions are assuming access to the rpm package which is not available is done in a bindist.
A possible solution would be to install the scripts as part of the package in a directory like `/usr/share/<pkg>/<script>.sh` so that they are available properly in the pkg_* phases. But that still doesn't take care of preinst...

**Comment by AndrewAmmerlaan** [comment]
Not sure if mirror restriction is needed but bindist is for sure. The pkg_* functions are assuming access to the rpm package which is not available is done in a bindist.

The mirror restriction is necessary because of the license, but that can stay in the ebuilds and doesn't have to go in the eclass I guess. Is there a way to force portage to download the contents of SRC_URI even if it is a binpkg? Otherwise I'll just add RESTRICT="bindist" to the eclass, which should prevent the problem from occurring. 

> A possible solution would be to install the scripts as part of the package in a directory like /usr/share/<pkg>/<script>.sh so that they are available properly in the pkg_* phases. But that still doesn't take care of preinst...

I could for example extract the scripts in pkg_unpack, and save them in some directory e.g. ${S}/var/tmp/${PF}/scripts/, this way they will be available if the package is a binpkg, and they would also be available in pkg_preinst, since pkg_unpack is run before pkg_preinst. 

Another solution would be to extract the scripts in the ebuild generation script and save them in ${FILESDIR}, but I'm pretty sure NI's license also covers the rpm-scripts meaning that we cannot redistribute those scripts in the repo. So probably this solution is legally impossible.

What do you think? add RESTRICT="bindist", or move the extraction of the scripts to the pkg_unpack phase and install them somewhere in ${S}/ (which will then later be installed to ${ED} and thus to the system)? I'm thinking the later will be less error-prone, but the disadvantage is that the script are installed to the file system.

**Comment by AndrewAmmerlaan** [comment]
> What do you think? add RESTRICT="bindist", or move the extraction of the scripts to the pkg_unpack phase and install them somewhere in ${S}/ (which will then later be installed to ${ED} and thus to the system)? I'm thinking the later will be less error-prone, but the disadvantage is that the script are installed to the file system.

Alright, I did some testing and it seems the latter solution is impossible. The DISTDIR does not exist in `pkg_setup`, and I can't use portageq to get it because emerge won't let me do that, which means I can't access the rpm file in this phase. I can't access the rpm command in the other phases, because it wants to write a lock file to /var/lock which results in ACCESS DENIED. The earliest I can do this is the pkg_preinst phase, but that is run *after* a hypothetical binpkg has been generated, so it does not solve our problem. It looks like we'll just have to RESTRICT="bindist" to solve the problem, which is not that bad of a solution since making a binary of the binary makes no sense to begin with.

**Comment by epsilon-0** [comment]
OMG OMG OMG, mistake, mistake, undo undo undo

**Pull request closed without merging**